### PR TITLE
Add rust example

### PR DESCRIPTION
**Pull request by gngeorgiev**
The Rust example is modeled as close as possible to the Go example.

### CODE REVIEW

File: rust/src/model/mod.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            address: "home".into(),
+            name: random_string(8),
+        }
+    }
+}
+
+#[derive(
+    Debug, Default, Clone, ValueList, FromRow, Serialize, StructFieldNames, FieldNamesAsArray,
+)]
+pub struct Pet {
```
Comment (gngeorgiev) [inline rust/src/model/mod.rs:38]: When the database is seeded, empty string values exist for pets, the driver seems to be unable to deserialize them back into a struct that has either an `Option<String>` or just a `String`. Not sure what the best way to handle that is since it's a pretty fundamental issue unless I messed it up somewhere.

**Comment by tzach** [comment]
@guy9 who can help us review the rust part?

**Comment by guy9** [comment]
@psarna can you or someone from your team review?

**Comment by psarna** [comment]
Sure, we'll review everything, but pretty much everyone is on vacation until 2022

File: rust/src/model/mod.rs (hunk 1 above)
Comment (psarna) [reply rust/src/model/mod.rs:38]: I did a few local tests and a struct with `Option<String>` worked just fine in accepting both nulls and empty strings, as expected.

However, I think I realize what the problem is. The `FromRow` macro does not match the names of attributes, but instead it works **only** on the order of fields.

So, given that the table is defined like that:
```cql
CREATE TABLE IF NOT EXISTS carepet.pet
(
    owner_id UUID,
    pet_id   UUID,
    chip_id  TEXT,
    species  TEXT,
    breed    TEXT,
    color    TEXT,
    gender   TEXT,
    age      INT,
    weight   FLOAT,
    address  TEXT,
    name     TEXT,
    PRIMARY KEY (owner_id, pet_id)
);
```
, that's the same order you should use in the struct in order for it to be deserialized properly:

```rust
    pub owner_id: Uuid,
    pub pet_id: Uuid,
    pub chip_id: Option<String>,
    pub species: Option<String>,
    pub breed: Option<String>,
    pub color: Option<String>,
    pub gender: Option<String>,
    pub age: i32,
    pub weight: f32,
    pub address: String,
    pub name: String,
```

@gngeorgiev please give it a try and let me know if it works.

Ref: https://github.com/scylladb/scylla-rust-driver/issues/370

Comment (gngeorgiev) [reply rust/src/model/mod.rs:38]: @psarna Thank you for looking into it. Initially, I had the fields ordered in the order you suggested, I moved the species field to be on top to make it more obvious that it's the offending field as I was getting an error for the field with index 3, which is the error I am getting now, after modifying the struct back to its original order. I have commited the new struct order but as mentioned the error persists. Please try following the instructions bellow, I hope that will help you reproduce it:

1. `docker run   -p 9042:9042/tcp   --name scylla   --hostname scylla   -d scylladb/scylla:latest    --smp 1`
2. `rustc 1.57.0 (f1edd0429 2021-11-29)`
3. 
```cql

CREATE KEYSPACE carepet WITH replication = {'class': 'NetworkTopologyStrategy', 'datacenter1': '1'}  AND durable_writes = true;

CREATE TABLE carepet.owner (
    owner_id uuid PRIMARY KEY,
    address text,
    name text
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE carepet.pet (
    owner_id uuid,
    pet_id uuid,
    address text,
    age int,
    breed text,
    chip_id text,
    color text,
    gender text,
    name text,
    species text,
    weight float,
    PRIMARY KEY (owner_id, pet_id)
) WITH CLUSTERING ORDER BY (pet_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE carepet.sensor (
    pet_id uuid,
    sensor_id uuid,
    type text,
    PRIMARY KEY (pet_id, sensor_id)
) WITH CLUSTERING ORDER BY (sensor_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE carepet.sensor_avg (
    sensor_id uuid,
    date date,
    hour int,
    value float,
    PRIMARY KEY (sensor_id, date, hour)
) WITH CLUSTERING ORDER BY (date ASC, hour ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE carepet.measurement (
    sensor_id uuid,
    ts timestamp,
    value float,
    PRIMARY KEY (sensor_id, ts)
) WITH CLUSTERING ORDER BY (ts ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';


INSERT INTO carepet.owner (owner_id, address, name) VALUES (322175a4-9891-4f30-84e0-767361880cd0, 'home', 'vcdpirmb');
INSERT INTO carepet.pet (owner_id, pet_id, address, age, breed, chip_id, color, gender, name, species, weight) VALUES (322175a4-9891-4f30-84e0-767361880cd0, 6b10cefb-2f09-4df3-8659-86e5588f9dad, 'home', 6, null, null, null, null, 'ouethhzv', null, 7.75718);
```
4. `cqlsh -f db.cql`
5. `cargo run`
6. `curl --request GET \
  --url http://127.0.0.1:8000/api/owner/322175a4-9891-4f30-84e0-767361880cd0/pets`

You should get: `{"error":"Bad CQL type in the column with index 3"}`

Comment (psarna) [reply rust/src/model/mod.rs:38]: Ugh, it's still about the order, it's just much more confusing than I initially suspected... let me explain more thoroughly. The way Cassandra (and Scylla) treats the order of columns is as follows:
1. partition key and clustering key are always created in order they appeared in the definition
2. static columns go next
3. regular columns come last, in the **lexicographical** order, regardless of the order in which they appeared in the CREATE TABLE statement.

You can see that by running a `DESCRIBE` command in cqlsh. Here's the output for the `pet` table:
```cql
cassandra@cqlsh> describe carepet.pet;

CREATE TABLE carepet.pet (
    owner_id uuid,
    pet_id uuid,
    address text,
    age int,
    breed text,
    chip_id text,
    color text,
    gender text,
    name text,
    species text,
    weight float,
    PRIMARY KEY (owner_id, pet_id)
) WITH CLUSTERING ORDER BY (pet_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

So the actual order for which your code will work (I verified this time, because your reproducing steps worked, thanks!) is as follows - I present it below in form of git diff:
```diff
diff --git a/rust/src/model/mod.rs b/rust/src/model/mod.rs
index 193f91d..d625edb 100644
--- a/rust/src/model/mod.rs
+++ b/rust/src/model/mod.rs
@@ -38,15 +38,15 @@ impl Owner {
 pub struct Pet {
     pub owner_id: Uuid,
     pub pet_id: Uuid,
-    pub chip_id: Option<String>,
-    pub species: Option<String>,
+    pub address: String,
+    pub age: i32,
     pub breed: Option<String>,
+    pub chip_id: Option<String>,
     pub color: Option<String>,
     pub gender: Option<String>,
-    pub age: i32,
-    pub weight: f32,
-    pub address: String,
     pub name: String,
+    pub species: Option<String>,
+    pub weight: f32,
 }
 
 impl Pet {
```

I agree that it's beyond confusing and it makes `FromRow` structs quite hard to grasp and use correctly - perhaps we should indeed consider manufacturing another macro which matches column names to attribute names, to avoid further confusion. /cc @piodul

Comment (psarna) [reply rust/src/model/mod.rs:38]: Alternatively, if you explicitly specify all the columns in the query (i.e. `SELECT owner_id, pet_id, chip_id, species, (...)` instead of just `SELECT *`), the values will be fetched in the order that you specified

Comment (gngeorgiev) [reply rust/src/model/mod.rs:55]: Thanks @psarna, this did fix it 🙂

Comment (gngeorgiev) [reply rust/src/model/mod.rs:55]: To clarify, I just changed the SELECTs to be explicit, since depending on the ordering of the fields is far from optimal IMO and is not future proof.

Comment (psarna) [reply rust/src/model/mod.rs:55]: I agree, an explicit select sounds better

**Comment by tarzanek** [comment]
also check https://docs.docker.com/desktop/mac/networking/#per-container-ip-addressing-is-not-possible
and recent changes to go docker approach (they integrated the app as part of scylla service to get network access even with above restriction) - https://github.com/scylladb/care-pet/pull/50
otherwise folks on Macs won't be able to run the example out of box (since their docker containers won't be directly accessible)

File: rust/src/bin/loadtest/main.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    care_pet::log::init();
+
+    let app: App = App::from_args();
+    debug!("Configuration = {:?}", app);
+
+    info!("Welcome to the Pets simulator");
+    rlimit::utils::increase_nofile_limit(102400).unwrap();
```
Comment (psarna) [inline rust/src/bin/loadtest/main.rs:53]: Panicking sounds a little harsh if you're not able to bump the file limit. Do you know why it needs to be bumped? Is it related to the number of connections? Or perhaps it just was there in the reference implementation in Go? Perhaps it's worth checking if the program also works without the bump, or just make it best-effort - try to increase the limit, but emit a warning and continue if it cannot be done.

File: rust/src/bin/loadtest/main.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    ));
+
+    info!("Creating flock");
+    let f = flock::new(app.owners, app.pets, app.sensors);
+
+    flock::save(sess, &f).await?;
+
+    let workers = app.workers.unwrap_or_else(num_cpus::get);
+    if app.writer {
+        let sensors = Arc::new(f.2);
```
Comment (psarna) [inline rust/src/bin/loadtest/main.rs:69]: I think that this line would become much more human-readable if Flock was upgraded to a regular struct with named fields instead of a tuple. That way the readers don't have to jump back to the definition if they want to check what's under `f.2`, if they instead see `f.sensors`.

File: rust/src/bin/loadtest/main.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    } else {
+        pets::start(sess, app.interval, f);
+        info!("Flock started");
+    }
+
+    // the application can only be terminated with interrupt signals, nothing to wait for
+    // specifically, so waiting for a channel that will never receive a value is fine
+    let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+    std::mem::forget(tx);
+    rx.await.map_err(From::from)
```
Comment (psarna) [inline rust/src/bin/loadtest/main.rs:87]: That snippet looks quite complicated. I think it's an equivalent of
```rust
let () = std::future::pending().await;
```
, which is quite self-descriptive and doesn't need the mem::forget trick.

File: rust/src/bin/loadtest/worker.rs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+use care_pet::{insert_query, random_sensor_data, Measure, Sensor};
+
+pub async fn work(sess: &'static Session, id: usize, sensors: Arc<HashMap<Uuid, Vec<Sensor>>>) {
+    debug!("worker # {} ready", id);
+
+    let prefix = format!("#{}", id);
+    let mut s = Stats::new();
+
+    loop {
+        for (_, sensors) in sensors.iter() {
```
Comment (psarna) [inline rust/src/bin/loadtest/worker.rs:21]: HashMap also implements Values, which gives you an iterator over values and would simplify the code:
```rust
for sensors in sensors.values() {
```

File: rust/src/db/keyspace.cql [hunk 6]
```diff
@@ -0,0 +1 @@
+CREATE KEYSPACE IF NOT EXISTS carepet WITH replication = { 'class': 'NetworkTopologyStrategy', 'replication_factor': '3' } AND durable_writes = TRUE;
```
Comment (psarna) [inline rust/src/db/keyspace.cql:1]: nitpick: durable_writes defaults to true and it's not going to change ever (it's crucial for Scylla/Cassandra to be fault-tolerant), so I'm pretty sure you can safely drop it from the keyspace definition. It doesn't hurt though

**Review by psarna — COMMENTED** [review]
I'm halfway through the review, but I'm leaving some comments in order not to block the progress

**Comment by tzach** [comment]
@psarna are you good with the latest updates?

**Comment by psarna** [comment]
I am - I also haven't gone through the whole PR yet, but the things I left for later are mostly about serialization, the REST API, etc. So from my perspective, it sounds already good to go if we're in a hurry (:

File: go/cmd/loadtest/flock.go [hunk 7]
```diff
@@ -120,6 +120,7 @@ func save(ctx context.Context, ses gocqlx.Session, f *flock) error {
 
 func startPets(ctx context.Context, f *flock, ses gocqlx.Session) {
 	for id := range f.Pets {
+		id := id
```
Comment (tzach) [inline go/cmd/loadtest/flock.go:123]: Its better to keep the Go update to another PR

**Comment by tzach** [comment]
OK, lets dont let this PR rot.
I'm merging it, and we can fix in issue in a followup PR
